// ktoolkit/src/http/user-agents.js
'use strict';

const core = {
    misc: require('../core/misc')
};

const UserAgentParts = {
    os: [
        'Windows NT 10.0; Win64; x64',
        'Windows NT 6.3; Win64; x64',
        'Windows NT 6.2; Win64; x64',
        'Windows NT 6.1; Win64; x64',
        'Windows NT 6.0; Win64; x64',
        'Windows NT 5.2; Win64; x64',
        'Windows NT 10.0; WOW64',
        'Windows NT 6.3; WOW64',
        'Windows NT 6.2; WOW64',
        'Windows NT 6.1; WOW64',
        'Windows NT 6.0; WOW64',
        'Windows NT 5.2; WOW64',
        'X11; Linux i686',
        'X11; Linux x86_64',
        'X11; FreeBSD i386',
        'X11; FreeBSD AMD64',
        'Macintosh; Intel Mac OS X'
    ],
    kernel: [{
        kernel: 'Trident/7.0',
        rv: '11.0',
        suffix: 'like Gecko'
    }, {
        kernel: 'Trident/7.0',
        rv: '10.0',
        suffix: 'like Gecko'
    }, {
        kernel: 'Trident/7.0',
        rv: '9.0',
        suffix: 'like Gecko'
    }, {
        rv: '50.0',
        suffix: 'Gecko/20100101 Firefox/50.0'
    }, {
        rv: '49.0',
        suffix: 'Gecko/20100101 Firefox/49.0'
    }, {
        rv: '48.0',
        suffix: 'Gecko/20100101 Firefox/48.0'
    }, {
        rv: '47.0',
        suffix: 'Gecko/20100101 Firefox/47.0'
    }, {
        rv: '46.0',
        suffix: 'Gecko/20100101 Firefox/46.0'
    }, {
        rv: '45.0',
        suffix: 'Gecko/20100101 Firefox/45.0'
    }, {
        rv: '44.0',
        suffix: 'Gecko/20100101 Firefox/44.0'
    }, {
        rv: '43.0',
        suffix: 'Gecko/20100101 Firefox/43.0'
    }, {
        rv: '42.0',
        suffix: 'Gecko/20100101 Firefox/42.0'
    }, {
        rv: '41.0',
        suffix: 'Gecko/20100101 Firefox/41.0'
    }, {
        rv: '40.0',
        suffix: 'Gecko/20100101 Firefox/40.0'
    }, {
        suffix: 'AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.106 Safari/537.36'
    }]
};

const UserAgents = [
    'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.106 Safari/537.36',
    'Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:50.0) Gecko/20100101 Firefox/50.0',
    'Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; rv:11.0) like Gecko'
];

function selectUserAgent() {
    let osIndex = core.misc.random(UserAgentParts.os.length);
    let kernelIndex = core.misc.random(UserAgentParts.kernel.length);

    let os = UserAgentParts.os[osIndex];
    let kernel = UserAgentParts.kernel[kernelIndex];

    let terms = [ os ];
    if ( kernel.kernel ) {
        terms.push(kernel.kernel);
    }
    if ( kernel.rv ) {
        terms.push(`rv: ${kernel.rv}`);
    }

    return `Mozilla/5.0 (${terms.join('; ')})${kernel.suffix}`;
}

module.exports = {
    selectUserAgent
};
